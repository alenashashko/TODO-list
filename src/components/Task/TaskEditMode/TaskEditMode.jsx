import React, {useState, useCallback} from "react";

export const TaskEditMode = (props) => {
  const {task, onModeChange, onTaskChange} = props;
  const {date, text, id} = task;

  const [taskText, setTaskText] = useState(text);

  const handleTaskTextChange = useCallback((evt) => {
    setTaskText(evt.target.value);
  }, []);

  const handleSaveEditing = useCallback(() => {
    onTaskChange(id, taskText);
    onModeChange("view");
  }, [onTaskChange, onModeChange, id, taskText]);

  const handleCancelEditing = useCallback(() => {
    onModeChange("view");
  }, [onModeChange]);

  return ( // блокировать кнопку save пока текст не изменен
    <div className="task-edit">
      <span>{date}</span>
      <textarea // отдельный компонент ? форма ?
        className="task-edit__text"
        name="task"
        placeholder="Start typing your text here..."
        value={taskText}
        onChange={handleTaskTextChange}
      />
      <button className="task-edit__save" onClick={handleSaveEditing}>Save</button>
      <button className="task-edit__cancel" onClick={handleCancelEditing}>Cancel</button>
    </div>
  );
};
