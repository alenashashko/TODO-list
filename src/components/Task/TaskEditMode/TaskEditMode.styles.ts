import { createStyles, makeStyles, Theme } from "@material-ui/core";

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    task: {
      display: "flex",
      flexDirection: "column",
      height: "100%",
    },
    content: {
      paddingTop: `${theme.spacing(3)}px`,
      paddingBottom: `${theme.spacing(3)}px`,
    },
    text: {
      width: "100%",

      "& textarea": {
        lineHeight: `${theme.spacing(6)}px`,
      },

      "& .MuiInputBase-multiline": {
        padding: `${theme.spacing(1)}px 0`,
      },
    },
    footer: {
      marginTop: "auto",
    },
    footerContainer: {
      display: "flex",
      alignItems: "flex-end",

      width: "100%",
      padding: `${theme.spacing(2)}px`,
    },
    buttons: {
      marginLeft: "auto",
    },
  })
);
