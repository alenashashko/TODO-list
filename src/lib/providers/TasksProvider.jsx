import React, {useContext, useState, useCallback, useMemo} from "react";
import {nanoid} from "nanoid";

const TasksContext = React.createContext({
  tasks: []
});

export const useTasksInfo = () => {
  const value = useContext(TasksContext);

  if (value === undefined) {
    throw new Error("you should use useTasksInfo inside of TasksProvider");
  }

  return value;
};

export const TasksProvider = (props) => {
  const {children} = props;

  const [tasks, setTasks] = useState([]);

  const createTask = useCallback((text) => {
    setTasks([...tasks, {
      id: nanoid(),
      date: Date.now(),
      text
    }])
  }, [tasks]);

  const changeTask = useCallback((taskId, newText) => {
    setTasks(tasks.map((task) => {
      return task.id !== taskId
        ? task
        : {
          ...task,
          date: Date.now(),
          text: newText
        }
    }))
  }, [tasks]);

  const deleteTask = useCallback((taskId) => {
    setTasks(tasks.filter((task) => task.id !== taskId));
  }, [tasks]);

  const value = useMemo(() => ({
    tasks,
    createTask,
    changeTask,
    deleteTask
  }), [tasks, createTask, changeTask, deleteTask]);

  return <TasksContext.Provider value={value}>{children}</TasksContext.Provider>;
};
