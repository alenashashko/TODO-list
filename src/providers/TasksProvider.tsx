import {FC, createContext, useContext, useState, useCallback, useMemo} from "react";
import {nanoid} from "nanoid";
import { Task } from "components/Task/Task";

const TasksContext = createContext<{
  tasks: Task[];
  createTask: (text: string) => void;
  changeTask: (taskId: string, newText: string) => void;
  deleteTask: (taskId: string) => void;
}>({
  tasks: [],
  createTask: () => {},
  changeTask: () => {},
  deleteTask: () => {},
});

export const useTasksInfo = () => {
  const value = useContext(TasksContext);

  if (value === undefined) {
    throw new Error("you should use useTasksInfo inside of TasksProvider");
  }

  return value;
};

export const TasksProvider: FC = (props) => {
  const {children} = props;

  const [tasks, setTasks] = useState<Task[]>([]);

  const createTask = useCallback((text: string) => {
    setTasks([...tasks, {
      id: nanoid(),
      date: Date.now(),
      text
    }])
  }, [tasks]);

  const changeTask = useCallback((taskId: string, newText: string) => {
    setTasks(tasks.map((task) => {
      return task.id !== taskId
        ? task
        : {
          ...task,
          date: Date.now(),
          text: newText
        }
    }))
  }, [tasks]);

  const deleteTask = useCallback((taskId: string) => {
    setTasks(tasks.filter((task) => task.id !== taskId));
  }, [tasks]);

  const value = useMemo(() => ({
    tasks,
    createTask,
    changeTask,
    deleteTask
  }), [tasks, createTask, changeTask, deleteTask]);

  return <TasksContext.Provider value={value}>{children}</TasksContext.Provider>;
};
