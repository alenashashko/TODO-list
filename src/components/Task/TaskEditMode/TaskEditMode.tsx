import { FC, useState, useCallback, ChangeEvent } from "react";
import {
  Card,
  CardActions,
  CardContent,
  Typography,
  TextField,
  IconButton,
  Tooltip,
} from "@material-ui/core";
import { Save as SaveIcon, Cancel as CancelIcon } from "@material-ui/icons";

import { useTasksInfo } from "src/providers/TasksProvider";
import { formatDate } from "src/utils/dates";
import { Task } from "src/components/Task/Task";

import { useStyles } from "./TaskEditMode.styles";

interface TaskEditModeProps {
  task: Task;
  onFinish: () => void;
}

export const TaskEditMode: FC<TaskEditModeProps> = ({ task, onFinish }) => {
  const { date, text, id } = task;

  const { changeTask } = useTasksInfo();
  const classes = useStyles();

  const [taskText, setTaskText] = useState<string>(text);

  const handleTaskTextChange = useCallback(
    (evt: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
      setTaskText((evt.target as HTMLTextAreaElement).value);
    },
    []
  );

  const handleSaveEditing = useCallback(() => {
    changeTask(id, taskText);
    onFinish();
  }, [changeTask, onFinish, id, taskText]);

  return (
    <Card className={classes.task}>
      <CardContent className={classes.content}>
        <TextField
          className={classes.text}
          name="task"
          placeholder="Start typing your text here..."
          multiline
          value={taskText}
          onChange={handleTaskTextChange}
        />
      </CardContent>
      <CardActions className={classes.footer}>
        <div className={classes.footerContainer}>
          <Typography color="textSecondary">{formatDate(date)}</Typography>
          <div className={classes.buttons}>
            <Tooltip title="Save" placement="top" arrow>
              <span>
                <IconButton
                  data-testid="save"
                  onClick={handleSaveEditing}
                  color="primary"
                  size="small"
                  aria-label="save changes"
                  disabled={taskText === text || taskText.trim().length === 0}
                >
                  <SaveIcon fontSize="small" />
                </IconButton>
              </span>
            </Tooltip>
            <Tooltip title="Cancel" placement="top" arrow>
              <IconButton
                data-testid="cancel"
                onClick={onFinish}
                color="secondary"
                size="small"
                aria-label="cancel changes"
              >
                <CancelIcon fontSize="small" />
              </IconButton>
            </Tooltip>
          </div>
        </div>
      </CardActions>
    </Card>
  );
};
