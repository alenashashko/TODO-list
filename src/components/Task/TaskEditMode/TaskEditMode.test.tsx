import { render, cleanup, fireEvent } from "@testing-library/react";

import { TaskEditMode } from "./TaskEditMode";

import { useTasksInfo } from "../../../providers/TasksProvider";
import { taskMock, tasksContextValueMock } from "../../../tests/mocks/tasks";

jest.mock("../../../providers/TasksProvider");

afterEach(cleanup);

it("renders task edit mode", () => {
  const mockedUseTasksInfo = useTasksInfo as jest.MockedFunction<typeof useTasksInfo>;

  mockedUseTasksInfo.mockReturnValue(tasksContextValueMock);

  const { asFragment } = render(
    <TaskEditMode
      task={taskMock}
      onFinish={() => { }}
    />);

  expect(asFragment()).toMatchSnapshot();
});

it("should call changeTask with correct id and text", () => {
  const mockedUseTasksInfo = useTasksInfo as jest.MockedFunction<typeof useTasksInfo>;

  const mockedChangeTask = jest.fn();

  mockedUseTasksInfo.mockReturnValue({
    ...useTasksInfo(),
    changeTask: mockedChangeTask,
  });

  const { getByPlaceholderText, getByTestId } = render(<TaskEditMode
    task={taskMock}
    onFinish={() => { }}
  />);

  const input = getByPlaceholderText("Start typing your text here...");

  const testText = "Learn JavaScript";

  fireEvent.change(input, {
    target: {
      value: testText,
    },
  });

  const saveButton = getByTestId("save");

  fireEvent.click(saveButton);

  expect(mockedChangeTask).toHaveBeenCalledTimes(1);
  expect(mockedChangeTask).toHaveBeenLastCalledWith(taskMock.id, testText);
});

it("should call onFinish when saveButton clicked", () => {
  const mockedUseTasksInfo = useTasksInfo as jest.MockedFunction<typeof useTasksInfo>;

  mockedUseTasksInfo.mockReturnValue(tasksContextValueMock);

  const mockedFinishEditing = jest.fn();

  const { getByPlaceholderText, getByTestId } = render(<TaskEditMode
    task={taskMock}
    onFinish={mockedFinishEditing}
  />);

  const input = getByPlaceholderText("Start typing your text here...");
  const testText = "Learn JavaScript";

  fireEvent.change(input, {
    target: {
      value: testText
    },
  });

  const saveButton = getByTestId("save");

  fireEvent.click(saveButton);

  expect(mockedFinishEditing).toHaveBeenCalledTimes(1);
});

it("should call onFinish when cancelButton clicked", () => {
  const mockedUseTasksInfo = useTasksInfo as jest.MockedFunction<typeof useTasksInfo>;

  mockedUseTasksInfo.mockReturnValue(tasksContextValueMock);

  const mockedFinishEditing = jest.fn();

  const { getByTestId } = render(<TaskEditMode
    task={taskMock}
    onFinish={mockedFinishEditing}
  />);

  const cancelButton = getByTestId("cancel");

  fireEvent.click(cancelButton);

  expect(mockedFinishEditing).toHaveBeenCalledTimes(1);
});
