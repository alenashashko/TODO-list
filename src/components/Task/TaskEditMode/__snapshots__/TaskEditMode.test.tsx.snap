// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders task edit mode 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-task-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardContent-root makeStyles-content-2"
    >
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-text-3"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-multiline MuiInput-multiline"
        >
          <textarea
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline"
            name="task"
            placeholder="Start typing your text here..."
            rows="1"
            style="height: 0px; overflow: hidden;"
          >
            my task
          </textarea>
          <textarea
            aria-hidden="true"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline"
            readonly=""
            style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;"
            tabindex="-1"
          />
        </div>
      </div>
    </div>
    <div
      class="MuiCardActions-root makeStyles-footer-4 MuiCardActions-spacing"
    >
      <div
        class="makeStyles-footerContainer-5"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
        >
          03/09/2021 3:50 PM
        </p>
        <div
          class="makeStyles-buttons-6"
        >
          <span
            class=""
            title="Save"
          >
            <button
              aria-label="save changes"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled MuiIconButton-sizeSmall Mui-disabled"
              data-testid="save"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
                  />
                </svg>
              </span>
            </button>
          </span>
          <button
            aria-label="cancel changes"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeSmall"
            data-testid="cancel"
            tabindex="0"
            title="Cancel"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
